File: python/cachefib.py
#!/usr/bin/env python
# File: cachefib.py
# Name: D.Saravanan
# Date: 20/03/2023

""" Script to compute nth Fibonacci number """


def fibonacci(n: int) -> int:
    """compute nth fibonacci"""

    cache: dict[int, int] = {}

    if n < 3:
        return 1

    if n in cache:
        return cache[n]

    cache[n] = fibonacci(n - 1) + fibonacci(n - 2)
    return cache[n]


if __name__ == "__main__":
    fnum: int = fibonacci(30)
    print(f"The 30th Fibonacci number is {fnum}")

File: julia/cachefib.jl
#!/usr/bin/env julia
# File: cachefib.jl
# Name: D.Saravanan
# Date: 20/03/2023

""" Script to compute nth Fibonacci number """


function fibonacci(n::Int, cache)::BigInt
    """compute nth fibonacci"""

    if n in keys(cache)
        return cache[n]
    end

    cache[n] = fibonacci(n - 1, cache) + fibonacci(n - 2, cache)
    return cache[n]
end


function main()
    cache = Dict{Int,BigInt}(1 => 1, 2 => 1)
    nval::Int = 1000
    fnum::BigInt = fibonacci(nval, cache)
    println("The $(nval)th Fibonacci number is $fnum")
end

main()
